### PR TITLE
MODFIN-71 Create transaction APIs - Payments, Credits

### PR DESCRIPTION
**Pull request by javalkar-varun**
## Purpose

https://issues.folio.org/browse/MODFIN-71

## Approach

- Implement Proxy to POST /finance-storage/invoice-transaction-summaries
- Implement Proxy to POST /finance-storage/transactions
- Implement Proxy to POST /finance-storage/transactions
- Update RMB version
- Add permissions
- Add unit tests

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/rest/helper/TransactionSummariesHelper.java [hunk 1]
```diff
@@ -19,13 +21,20 @@ public TransactionSummariesHelper(Map<String, String> okapiHeaders, Context ctx,
...
 
-  private void validateTransactionCount(OrderTransactionSummary orderSummary) {
-    if (orderSummary.getNumTransactions() < 1) {
+  public CompletableFuture<InvoiceTransactionSummary> createInvoiceTransactionSummary(InvoiceTransactionSummary invoiceSummary) {
+    return VertxCompletableFuture.runAsync(ctx, () -> validateTransactionCount(null, invoiceSummary))
+      .thenCompose(ok -> handleCreateRequest(resourcesPath(INVOICE_TRANSACTION_SUMMARIES), invoiceSummary))
+      .thenApply(invoiceSummary::withId);
+  }
+
+  private void validateTransactionCount(OrderTransactionSummary orderSummary, InvoiceTransactionSummary invoiceSummary) {
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/helper/TransactionSummariesHelper.java]: The method will become simpler if it has one argument of integer type.

File: src/main/java/org/folio/rest/impl/TransactionSummariesAPI.java [hunk 2]
```diff
@@ -30,5 +32,15 @@ public void postFinanceOrderTransactionSummaries(String lang, OrderTransactionSu
       .exceptionally(fail -> handleErrorResponse(asyncResultHandler, helper, fail));
   }
 
+  @Override
+  @Validate
+  public void postFinanceInvoiceTransactionSummaries(String lang, InvoiceTransactionSummary entity,
+      Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    TransactionSummariesHelper helper = new TransactionSummariesHelper(okapiHeaders, vertxContext, lang);
+    helper.createInvoiceTransactionSummary(entity)
+      .thenAccept(type -> asyncResultHandler.handle(succeededFuture(helper.buildResponseWithLocation(String.format(INVOICE_TRANSACTION_SUMMARIES_LOCATION_PREFIX, type.getId()), type))))
```
Comment (grabsefx) [inline src/main/java/org/folio/rest/impl/TransactionSummariesAPI.java]: Please, rename parameter name "type" to "txSummary" or "transactionSummary".

File: src/main/java/org/folio/rest/helper/TransactionSummariesHelper.java [hunk 3]
```diff
@@ -19,14 +24,25 @@ public TransactionSummariesHelper(Map<String, String> okapiHeaders, Context ctx,
...
+  public CompletableFuture<InvoiceTransactionSummary> createInvoiceTransactionSummary(InvoiceTransactionSummary invoiceSummary) {
+    return VertxCompletableFuture
+      .runAsync(ctx,
+          () -> validateTransactionCount(
+              Arrays.asList(invoiceSummary.getNumPaymentsCredits(), invoiceSummary.getNumEncumbrances())))
+      .thenCompose(ok -> handleCreateRequest(resourcesPath(INVOICE_TRANSACTION_SUMMARIES), invoiceSummary))
+      .thenApply(invoiceSummary::withId);
+  }
+
+  private void validateTransactionCount(List<Integer> summaryDetailCounts) {
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/helper/TransactionSummariesHelper.java]: The method will become simpler and easier to use if it has one argument of **integer** type.

File: src/main/java/org/folio/rest/impl/TransactionsApi.java [hunk 4]
```diff
@@ -93,4 +93,37 @@ public void getFinanceTransactionsById(String id, String lang, Map<String, Strin
       .exceptionally(fail -> handleErrorResponse(asyncResultHandler, helper, fail));
   }
 
+  @Validate
+  @Override
+  public void postFinancePayments(String lang, Transaction payment, Map<String, String> okapiHeaders,
+      Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
+    TransactionsHelper helper = new TransactionsHelper(okapiHeaders, vertxContext, lang);
+
+    if (payment.getTransactionType() != Transaction.TransactionType.PAYMENT) {
```
Comment (pautau-s) [inline src/main/java/org/folio/rest/impl/TransactionsApi.java]: This code is repeated in many places. Could you extract it into a method?

File: src/main/java/org/folio/rest/helper/TransactionSummariesHelper.java (hunk 3 above)
Comment (javalkar-varun) [reply src/main/java/org/folio/rest/helper/TransactionSummariesHelper.java]: Currently, there only 2 methods using it. invoiceSummary has 2 counts. Are you suggesting something like `validateTransactionCount(Integer summaryCounts1, Integer summaryCounts2)` and call it like `validateTransactionCount(orderSummary.getNumTransactions(), null)` ?

Comment (pautau-s) [reply src/main/java/org/folio/rest/helper/TransactionSummariesHelper.java]: I'm suggesting something like `validateTransactionCount(Integer summaryCounts)`. And then call it  like:
`validateTransactionCount(invoiceSummary.getNumPaymentsCredits())`, `validateTransactionCount(invoiceSummary.getNumEncumbrances())` or `validateTransactionCount(orderSummary.getNumTransactions())`

**Pull request merged**